File: rust/2015/day20a/src/main.rs
use std::env;
use std::io;
use std::process;

fn usage(progname: &str) {
    eprintln!("usage: {progname} <num>");
    process::exit(1);
}

fn process_input(s: &str) -> io::Result<u32> {
    let n = s.parse::<u32>().expect("input must be a number!");
    let limit = n / 10;
    let mut houses: Vec<u32> = vec![0; (limit + 1) as usize];
    for elf in 1..=limit {
        let gift = elf * 10;
        for house in (elf..=limit).step_by(elf as usize) {
            houses[house as usize] += gift;
        }
    }
    let mut i = 1;
    loop {
        let p = houses[i];
        if p >= n {
            return Ok(i as u32);
        } else {
            i += 1;
        }
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let progname = &args[0];
    if args.len() < 2 {
        usage(progname);
    }
    let s = &args[1];
    match process_input(s) {
        Ok(result) => println!("result = {result}"),
        Err(e) => {
            eprintln!("error while processing input: {e}");
            process::exit(1);
        }
    }
}

File: haskell/2015/day20a/src/Main.hs
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE FlexibleContexts #-}

module Main (main) where

import Control.Monad
import Control.Monad.ST
import Data.Array.ST
import Data.Int (Int32)
import System.Environment (getArgs, getProgName)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)
import Control.DeepSeq
import System.Clock

usage :: String -> IO ()
usage progname = do
    hPutStrLn stderr $ "usage: " ++ progname ++ " <n>"
    exitFailure

solve :: Int -> Int
solve n = runST $ do
    let limit = n `div` 10
    arr <- newArray (1, limit) 0 :: ST s (STUArray s Int Int)
    forM_ [1..limit] $ \elf -> do
        let gift = elf * 10
        forM_ [elf, 2*elf .. limit] $ \house -> do
            cur <- readArray arr house
            writeArray arr house (cur + gift)
    let go !i
          | i > limit = error "unreachable"
          | otherwise = do
              p <- readArray arr i
              if p >= n then return i else go (i + 1)
    go 1

process :: Int -> Int32
process = fromIntegral . solve


showTime :: TimeSpec -> String
showTime elapsed =
    let ns = fromIntegral (toNanoSecs elapsed) :: Double
    in if ns < 1000
       then show ns ++ " ns"
       else if ns < 1000000
       then show (ns / 1000.0) ++ " μs"
       else if ns < 1000000000
            then show (ns / 1000000.0) ++ " ms"
            else show (ns / 1000000000.0) ++ " s"
main :: IO ()
main = do
    args <- getArgs
    progname <- getProgName
    case args of
        [n] -> do
            start <- getTime Monotonic
            let result = process (read n)
            result `deepseq` return ()
            end <- getTime Monotonic
            let elapsed = diffTimeSpec start end
            putStrLn $ "result = " ++ show result
            putStrLn $ "elapsed time: " ++ showTime elapsed
        _ -> usage progname
